### ISSUE TITLE
Kafka producer settings may cause message reordering

### ISSUE THREAD

**Issue by wushujames** [zendesk/maxwell#241]
Ben,

I think there is potential for data loss or message reordering when using the Kafka producer. Have you seen this slide deck?

http://www.slideshare.net/JiangjieQin/no-data-loss-pipeline-with-apache-kafka-49753844

In particular, think of the following scenario:
1) Two messages in the binlog, message 1 and message 2
2) Maxwell does producer.send(message1); producer.send(message2);
3) Broker for message1 is delayed (and it's on a different partition than message2). Message2 gets delivered first.
4) Message2's onCompletion() gets called, which checkpoints message2.
5) Maxwell crash.

Upon startup, the binlog position is at message2. Message1 never got delivered.

The slidedeck also talks about possible message reordering.

Do you think the Kafka producer settings should be defaulted to the ones in the slide deck? I know that you also allow users to provide their own Kafka configuration in their config.properties file. Should this be a user-configurable setting, or should we default to the zero-data-loss settings?

**Comment by osheroff** [comment]
hmm, you're right.  I was misunderstanding the semantics of the `Producer`:

>Callbacks for records being sent to the same partition are guaranteed to execute in order. That is, in the following example callback1 is guaranteed to execute before callback2:
```
producer.send(new ProducerRecord<byte[],byte[]>(topic, partition, key1, value1), callback1);
producer.send(new ProducerRecord<byte[],byte[]>(topic, partition, key2, value2), callback2);
```

but I had thought that the callback would be serialized per *Producer*.

Now, it would probably kill our performance to only send one record at a time (not to mention it would hurt compression ratios), but we can do something where we maintain a list of in-flight kafka requests, and only write the binlog position up until where all the messages have been delivered, ie:

```
  kafka_inflight:  [12 / undone, 19 / undone, 23 / undone, 24 / undone]
```

now if a completion request comes in for 19, we mark it done, but don't checkpoint the binlog:
```
  kafka_inflight:  [12 / undone, 19 / done, 23 / undone, 24 / undone]
```

and when the completion callback for 12 comes in, we mark it done and dequeue 12 and 19, setting the binlog position to "19".

The other issues mentioned in that slidedeck are a bit muckier.  I'd be curious to see what impact `acks=all` has on performence.

**Comment by wushujames** [comment]
Yeah, I was thinking of something like you propose: only checkpoint a request when all previous ones have checkpointed. Or said another way, only checkpoint if you are at the beginning of the queue.

When you say "it would probably kill our performance to only send one record at a time", did you mean you want to avoid producer.send().get()? (synchronous sends)?

The message reordering issue is particularly important since you are doing database replication. You don't want 2 row updates (for the same primary key) to come out of order.

**Comment by gowthamgutha** [comment]
@osheroff Even if both message1 and message2 belong to the same partition and the max.in.flight.requests is set to a value >1 then is it not possible that the message2 gets delivered first than message1? (provided that the both belong to different batches).

Because, after all since both of the messages are sent independently over the network it is uncertain which reaches the broker first.

**Comment by osheroff** [comment]
@gowthamgutha if message1 and message2 are on the same partition they will be delivered over the same socket, and TCP will guarantee that the messages are received in-order.
